tests/kernel/mem_protect/mpu/src/mem_domain.c

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_mpu.h"
+
+#define ERROR_STR \
+	("Fault didn't occur when we accessed a unassigned memory domain.")
+
+/****************************************************************************/
+K_THREAD_STACK_DEFINE(mem_domain_1_stack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(mem_domain_2_stack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(mem_domain_6_stack, MEM_DOMAIN_STACK_SIZE);
+__kernel struct k_thread mem_domain_1_tid, mem_domain_2_tid, mem_domain_6_tid;
+
+/****************************************************************************/
+/* The mem domains needed.*/
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_buf[MEM_REGION_ALLOC];
+
+/* partitions added later in the test cases.*/
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part1[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part2[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part3[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part4[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part5[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part6[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part7[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT mem_domain_tc3_part8[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(mem_domain_memory_partition,
+		       mem_domain_buf,
+		       sizeof(mem_domain_buf),
+		       K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *mem_domain_memory_partition_array[] = {
+	&mem_domain_memory_partition
+};
+
+struct k_mem_domain mem_domain_mem_domain;
+
+/****************************************************************************/
+/* Common init functions */
+static inline void mem_domain_init(void)
+{
+	k_mem_domain_init(&mem_domain_mem_domain,
+			  MEM_PARTITION_INIT_NUM,
+			  mem_domain_memory_partition_array);
+}
+
+static inline void set_valid_fault_value(int test_case_number)
+{
+	switch(test_case_number){
+	case 1:
+		valid_fault = false;
+		break;
+	case 2:
+		valid_fault = true;
+		break;
+	default:
+		valid_fault = false;
+		break;
+	}
+}
+
+/****************************************************************************/
+/* Userspace function */
+void mem_domain_for_user(void *tc_number, void *p2, void *p3)
+{
+	set_valid_fault_value((u32_t)tc_number);
+
+	mem_domain_buf[0]= 10;
+	if (valid_fault == false)
+		ztest_test_pass();
+}
```

Comment:
Might it be better to use two different test functions for these two different cases/expected results, thereby avoiding complicating the code and implication of a fault from the same function?